sync15-adapter/src/state_machine.rs

```diff
@@ -0,0 +1,599 @@
+                        scratchpad.save();
+                        Ok(HasMetaGlobal { collections })
+                    }
+                    MetaGlobalEnginesAction::ChangeStates(changes) => {
+                        scratchpad.set_global(Some(global));
+                        for change in changes {
+                            scratchpad.add_engine_state_change(change);
+                        }
+                        scratchpad.save();
+                        Ok(HasMetaGlobal { collections })
+                    }
+                }
+            }
+
+            // Check if our locally cached `crypto/keys` collection is
+            // up-to-date.
+            HasMetaGlobal { collections } => {
+                let action = match scratchpad.keys() {
+                    Some(keys) => match &collections.get("crypto") {
+                        Some(modified) => {
+                            if keys.modified >= **modified {
+                                FetchAction::Skip
+                            } else {
+                                FetchAction::Fetch
+                            }
+                        }
+                        None => FetchAction::InvalidateThenFetch,
+                    },
+                    None => FetchAction::Fetch,
+                };
+                Ok(match action {
+                    // If `crypto/keys` is up-to-date, we're ready to go!
+                    FetchAction::Skip => Ready,
+                    // We need to fetch and cache new keys.
+                    FetchAction::Fetch => NeedsFreshCryptoKeys,
+                    // We need to invalidate our locally cached `crypto/keys`,
+                    // then try to fetch new keys, and reupload if fetching
+                    // fails.
+                    FetchAction::InvalidateThenFetch => {
+                        scratchpad.set_keys(None);
+                        NeedsFreshCryptoKeys
+                    }
+                })
+            }
+
+            NeedsFreshCryptoKeys => {
+                match client.fetch_crypto_keys() {
+                    Ok(encrypted_bso) => {
+                        let keys = CollectionKeys::from_encrypted_bso(encrypted_bso, root_key)?;
+                        scratchpad.add_engine_state_changes_from_keys(Some(&keys));
+                        scratchpad.set_keys(Some(keys));
+                        Ok(Ready)
+                    }
+                    // If the server doesn't have a `crypto/keys`, start over
+                    // and reupload our `meta/global` and `crypto/keys`.
+                    Err(error::Error(error::ErrorKind::NoCryptoKeys, _)) => Ok(FreshStartRequired),
+                    Err(err) => Err(err),
+                }
+            }
+
+            Ready => bail!(error::unexpected("Can't advance past ready")),
+
+            FreshStartRequired => {
+                client.wipe_storage()?;
```

Comment:
I'm slightly uncomfortable with this doing wipe_storage, particularly for all collections when the consumer of this might not support all collections.